File: lib/domain/crop/policies/crop_masters_nested_access.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Policies
      # 作物マスタ配下 API（農薬・害虫等）: ユーザー所有の非参照作物への edit 可否。
      module CropMastersNestedAccess
        module_function

        def assert_edit_allowed_for_masters!(user, crop_entity)
          access_filter = Domain::Shared::Policies::CropPolicy.record_access_filter(user)
          Domain::Shared::ReferenceRecordAuthorization.assert_edit_allowed!(access_filter, crop_entity)
        end
      end
    end
  end
end

File: crates/agrr-domain/src/pesticide/policies/crop_masters_nested_access.rs
use crate::shared::policies::crop_policy;
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;
use crate::shared::record_ref::RecordRef;
use crate::shared::reference_record_authorization;
use crate::shared::user::User;

/// Ruby: `Domain::Crop::Policies::CropMastersNestedAccess`
pub fn assert_edit_allowed_for_masters<R: RecordRef>(
    user: User,
    crop_entity: &R,
) -> Result<(), PolicyPermissionDenied> {
    let access_filter = crop_policy::record_access_filter(user);
    reference_record_authorization::assert_edit_allowed(&access_filter, crop_entity)
}
